### PR TITLE
[FLINK-25033][runtime] Let some scheduler components updatable

### PR DESCRIPTION
**Pull request by wanglijie95**
## What is the purpose of the change

Let some scheduler components updatable, including DefaultExecutionTopology, SchedulingStrategy, PartitionReleaseStrategy and SlotSharingStrategy.

## Brief change log
 ca7544446ab6ca1aebf38c3b338bc5bdf3864d14 DefaultExecutionTopology supports update.
 9294b67a0f8f7b6881e581a41cf3895a8196ff98 RegionPartitionGroupReleaseStrategy and LocalInputPreferredSlotSharingStrategy support update.

## Verifying this change
unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/partitionrelease/RegionPartitionGroupReleaseStrategy.java [hunk 1]
```diff
@@ -144,6 +145,52 @@ private PipelinedRegionExecutionView getPipelinedRegionExecutionViewForVertex(
...
+                for (ConsumedPartitionGroup consumedPartitionGroup :
+                        pipelinedRegion.getAllBlockingConsumedPartitionGroups()) {
+                    partitionGroupConsumerRegions
+                            .computeIfAbsent(
+                                    consumedPartitionGroup,
+                                    g -> {
+                                        ConsumerRegionGroupExecutionView regionGroup =
+                                                new ConsumerRegionGroupExecutionView();
+                                        this.consumerRegionGroupExecutionViewMaintainer
+                                                .notifyNewRegionGroup(regionGroup);
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/partitionrelease/RegionPartitionGroupReleaseStrategy.java]: You are right, it's a bug, I fixed it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java [hunk 2]
```diff
@@ -59,8 +59,10 @@
 import org.apache.flink.runtime.jobgraph.JobVertex;
 import org.apache.flink.runtime.jobgraph.JobVertexID;
 import org.apache.flink.runtime.jobgraph.tasks.CheckpointCoordinatorConfiguration;
+import org.apache.flink.runtime.jobmanager.scheduler.SlotSharingGroup;
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java:62]: Looks to me commit 6cb03fe1f9824c011f752a416197d1e2004a2e11 (Support calcuate network memory for dynamic graph.) is not related to the topic of this PR "Let some scheduler components updatable"?
If not, would you create a separate ticket and PR to make this change?

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java [hunk 3]
```diff
@@ -159,50 +162,79 @@ public static DefaultExecutionTopology fromExecutionGraph(
...
+        newJobVertices.stream()
+                .map(ExecutionJobVertex::getJobVertex)
+                .flatMap(v -> v.getInputs().stream())
+                .map(JobEdge::getSource)
+                .filter(r -> r.getResultType().isPipelined())
+                .map(IntermediateDataSet::getProducer)
+                .map(JobVertex::getID)
+                .forEach(id -> checkState(newVertexSet.contains(id)));
+
+        final Iterable<ExecutionVertex> newAddedExecutionVertices =
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java:209]: newAddedExecutionVertices -> newlyAddedExecutionVertices or newExecutionVertices

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java [hunk 4]
```diff
@@ -159,50 +162,79 @@ public static DefaultExecutionTopology fromExecutionGraph(
...
+        return logicalPipelinedRegionsByJobVertexId;
+    }
+
+    public void notifyExecutionGraphUpdated(
+            final DefaultExecutionGraph executionGraph,
+            final List<ExecutionJobVertex> newJobVertices) {
+
+        checkNotNull(executionGraph, "execution graph can not be null");
+
+        final Set<JobVertexID> newVertexSet =
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java:193]: NIT: newVertexSet -> newJobVertexIds

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java [hunk 5]
```diff
@@ -159,50 +162,79 @@ public static DefaultExecutionTopology fromExecutionGraph(
...
+                executionGraph,
+                IterableUtils.toStream(executionGraph.getVerticesTopologically())
+                        .filter(ExecutionJobVertex::isInitialized)
+                        .collect(Collectors.toList()));
+
+        return schedulingTopology;
+    }
+
+    private void updateExecutionGraphIndex(Iterable<ExecutionVertex> executionVertices) {
+
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java:261]: NIT: unnecessary empty line

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java [hunk 6]
```diff
@@ -159,50 +162,79 @@ public static DefaultExecutionTopology fromExecutionGraph(
...
+        schedulingTopology.notifyExecutionGraphUpdated(
+                executionGraph,
+                IterableUtils.toStream(executionGraph.getVerticesTopologically())
+                        .filter(ExecutionJobVertex::isInitialized)
+                        .collect(Collectors.toList()));
+
+        return schedulingTopology;
+    }
+
+    private void updateExecutionGraphIndex(Iterable<ExecutionVertex> executionVertices) {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java:260]: executionVertices -> newExecutionVertices

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java [hunk 7]
```diff
@@ -283,13 +309,25 @@ private static DefaultExecutionVertex generateSchedulingExecutionVertex(
...
     }
 
-    private static IndexedPipelinedRegions computePipelinedRegions(
-            Iterable<DefaultLogicalPipelinedRegion> logicalPipelinedRegions,
-            Function<DefaultLogicalPipelinedRegion, List<DefaultExecutionVertex>>
-                    sortedExecutionVerticesInPipelinedRegion,
-            Function<ExecutionVertexID, DefaultExecutionVertex> executionVertexRetriever,
-            Function<IntermediateResultPartitionID, DefaultResultPartition>
-                    resultPartitionRetriever) {
+    private void updatePipelinedRegions(Iterable<ExecutionVertex> executionVertices) {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java:336]: executionVertices -> newExecutionVertices

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java (hunk 6 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java:260]: > updateExecutionGraphIndex

The name is outdated.
Maybe rename it to `generateNewExecutionVerticesAndResultPartitions`, or any other good ideas?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategyTest.java [hunk 8]
```diff
@@ -301,6 +302,57 @@ public void testInputLocalityIsRespectedWithTwoEdgesBetweenTwoVertices() throws
...
+        assertThat(
+                strategy.getExecutionSlotSharingGroup(ev1.getId()).getExecutionVertexIds(),
+                containsInAnyOrder(ev1.getId(), ev2.getId()));
+        assertThat(
+                strategy.getExecutionSlotSharingGroup(ev2.getId()).getExecutionVertexIds(),
+                containsInAnyOrder(ev1.getId(), ev2.getId()));
+
+        // add new job vertices and notify scheduling topology updated
+        TestingSchedulingExecutionVertex ev3 = topology.newExecutionVertex(jobVertexID3, 0);
+        topology.connect(ev2, ev3);
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategyTest.java:341]: -> topology.connect(ev2, ev3, ResultPartitionType.BLOCKING);

a.t.m we do not expect pipelined downstream vertices to be attached dynamically

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java (hunk 2 above)
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java:62]: I have separated the "Support calculate network memory for dynamic graph" and "Support register operator coordinators for new initialized ExecutionJobVertex" into two new tickets: FLINK-25668 and FLINK-25669

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultResultPartition.java [hunk 9]
```diff
@@ -28,12 +28,13 @@
...
 
 import java.util.List;
+import java.util.Optional;
 import java.util.function.Supplier;
 
 import static org.apache.flink.util.Preconditions.checkNotNull;
 
 /** Default implementation of {@link SchedulingResultPartition}. */
-class DefaultResultPartition implements SchedulingResultPartition {
+public class DefaultResultPartition implements SchedulingResultPartition {
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultResultPartition.java:37]: It's not needed, I will remove it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultResultPartition.java [hunk 10]
```diff
@@ -54,16 +55,32 @@
...
             Supplier<ResultPartitionState> resultPartitionStateSupplier,
-            ConsumerVertexGroup consumerVertexGroup,
+            Supplier<ConsumerVertexGroup> consumerVertexGroupSupplier,
             Supplier<List<ConsumedPartitionGroup>> consumerPartitionGroupSupplier) {
         this.resultPartitionId = checkNotNull(partitionId);
         this.intermediateDataSetId = checkNotNull(intermediateDataSetId);
         this.partitionType = checkNotNull(partitionType);
         this.resultPartitionStateSupplier = checkNotNull(resultPartitionStateSupplier);
-        this.consumerVertexGroup = consumerVertexGroup;
+        this.consumerVertexGroupSupplier = consumerVertexGroupSupplier;
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultResultPartition.java:64]: OK, I will add a hotfix commit to add `checkNotNull` for `Supplier/Function`

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java (hunk 6 above)
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java:260]: `generateNewExecutionVerticesAndResultPartitions ` is good for me, and I will also rename the`updatePipelinedRegions ` to `generateNewPipelinedRegions`

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/partitionrelease/RegionPartitionGroupReleaseStrategy.java [hunk 11]
```diff
@@ -144,6 +149,54 @@ private PipelinedRegionExecutionView getPipelinedRegionExecutionViewForVertex(
...
+        final PipelinedRegionExecutionView regionExecutionView =
+                getPipelinedRegionExecutionViewForVertex(executionVertexId);
+        return regionExecutionView.isFinished();
+    }
+
+    @Override
+    public void notifySchedulingTopologyUpdated(
+            SchedulingTopology schedulingTopology, List<ExecutionVertexID> newlyAddedVertices) {
+
+        final List<ConsumerRegionGroupExecutionView> newlyAddedRegionGroup = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/partitionrelease/RegionPartitionGroupReleaseStrategy.java:163]: OK, I will change the method accoridng your suggesstion.

**Comment by wanglijie95** [comment]
@zhuzhurk Thanks for review. I've addressed all the comments , please help to review it again.

**Review by zhuzhurk — APPROVED** [review]
Thanks for addressing all the comments! @wanglijie95 
The change now looks good to me.

**Comment by zhuzhurk** [comment]
Merging.

**Pull request closed without merging**